apps/emqx_authz/src/emqx_authz.erl

```diff
@@ -377,5 +380,30 @@ type(Unknown) -> error({unknown_authz_source_type, Unknown}).
 acl_conf_file() ->
     filename:join([emqx:data_dir(), "authz", "acl.conf"]).
 
+%% http resource url field encode
+http_fullpath_uri_encode(RawURL) ->
+    case cow_http:parse_fullpath(to_bin(RawURL)) of
+        {BaseURL, <<>>} ->
+            BaseURL;
+        {BaseURL, QS} ->
+            Decoded = emqx_http_lib:uri_decode(QS),
+            NQS = emqx_http_lib:uri_encode(Decoded),
+            <<BaseURL/binary, "?", (emqx_http_lib:uri_encode(NQS))/binary>>
+    end.
+
 ph_to_re(VarPH) ->
     re:replace(VarPH, "[\\$\\{\\}]", "\\\\&", [global, {return, list}]).
+
+ph_to_re_percent(PH) ->
+    Fun = fun ({RE, PercentEncode}, PhAccIn) ->
+                  re:replace(PhAccIn, RE, PercentEncode, [global, {return, binary}])
+          end,
+    %% "$" is not necessary to percent encode.
+    %% If does, it would be back to "$" from "%24" when decoding.
+    ReplaceTuples = [{"[\\$]", "\\\\&"}, {"[\\{]", "%7B"}, {"[\\}]", "%7D"}],
+    lists:foldl(Fun, PH, ReplaceTuples).
+
+to_bin(B) when is_binary(B) ->
+    B;
+to_bin(L) when is_list(L) ->
+    list_to_binary(L).
```

Comment:
his this only for authz ? or can be general abstraction for other http APIs too ?

Reply:
We can parse the RawURL `https://fake.com/test/path?username=${username}&clientid=${clientid}` by cowlib functinos.
Then got a query string KV map like
`#{<<"username">> => <<"${username}">>, <<"clientid">> => <<"${clientid}">>}`
It will be encode automactily when generating url to validate.

We have no need to percent encoding the placeholders in this way.
But refactoring may take some time base on #6506.